src/tls13.rs

```diff
@@ -0,0 +1,222 @@
+                );
+                return Err(rustls::Error::General(custom_error_message));
+            }
+        }
+    }
+}
+
+/// [`MessageDecrypter`] for ChaCha 1.3
+/// the [`payload`] field that comes from the [`OpaqueMessage`] is structured to include the message which is an arbitrary length, 
+/// an encoding type that is 1 byte and then finally the tag which is 16 bytes. 
+/// ex : [1, 2, 3, 5, 6, 7, 8, 9, 10, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12 ,13, 14, 15, 16]
+///       ^                       ^   ^  ^                                                   ^
+///            Message (N bytes)   Encoding (1 Byte)              Tag (16 bytes)
+impl MessageDecrypter for Tls13ChaCha20Poly1305 {
+    fn decrypt(&self, mut msg: rustls::crypto::cipher::OpaqueMessage, seq: u64) -> Result<rustls::crypto::cipher::PlainMessage, rustls::Error> {
+        let mut payload = msg.payload_mut();
+        let payload_len = payload.len(); // This length includes the message, encoding, and tag.
+
+        // Ensure that the length is over 16 bytes or there is a decryption error.
+        if payload_len < CHACHA_TAG_LENGTH {
+            return Err(rustls::Error::DecryptError);
+        }
+        let message_length = payload_len - CHACHA_TAG_LENGTH; // getting message length, this includes the message length and the encoding type.
+
+        // Set up needed parameters for ChaCha decrypt
+        let nonce = Nonce::new(&self.iv, seq);
+        let auth_data = make_tls13_aad(payload_len);
+        let mut tag = [0u8; 16];
+        tag.copy_from_slice(&payload[message_length..]);
+        
+        // Decrypting the payload in place, there is no +1 here since [`message_length`] accounts for the extra byte for encoding type.
+        match chacha20_poly1305_decrypt_in_place(&self.key, &nonce.0, &auth_data, &mut payload[..message_length], &tag) {
+            Ok(_) => {
+                payload.truncate(message_length);
+                msg.into_tls13_unpadded_message()
+            }
+            Err(symcrypt_error) => {
+                let custom_error_message = format!(
+                    "SymCryptError: {}",
+                    symcrypt_error.to_string() // Using general error to propagate the SymCrypt error back to the caller
+                );
+                return Err(rustls::Error::General(custom_error_message));
+            }
+        }
+    }
+}
+
+/// GCM for TLS 1.3
+
+/// Tls13Gcm impls Tls13AeadAlgorithm.
+pub struct Tls13Gcm;
+
+/// Tls13GcmState impls [`MessageEncrypter`] and [`MessageDecrypter`]
+/// 
+/// [`key`] is a rust-symcrypt::GcmExpandedKey that has expands the provided key
+/// [`iv`] is an initialization vector that is needed to create the unique nonce.
+pub struct Tls13GcmState {
+    key: GcmExpandedKey,
+    iv: Iv
+}
+
+impl Tls13AeadAlgorithm for Tls13Gcm { 
+    fn encrypter(&self, key: rustls::crypto::cipher::AeadKey, iv: Iv) -> Box<dyn MessageEncrypter> {

```

Comment:
>key: rustls::crypto::cipher::AeadKey [](http://example.com/codeflow?start=24&length=36)

Can this key have an invalid length?

Not sure if we want to panic for that?

Reply:
Add something like:
     assert_eq!(key.as_ref().len(), 32); // Gcm key length must be 32 bytes.

Reply:
It can be 16 or 32